Try: Remove segmented control vertical separators

jasmussen opened the pull request:
## Description

Inspired by https://github.com/WordPress/gutenberg/pull/35395#issuecomment-938615968, this PR removes the vertical separators from the Segmented Control entirely, as in most examples of its usage I have found them to only add noise.

Before:

<img width="1270" alt="before" src="https://user-images.githubusercontent.com/1204802/136757866-dc51debd-2f8a-4a9d-894a-d470decc232e.png">

After:

<img width="1270" alt="after" src="https://user-images.githubusercontent.com/1204802/136757882-29fe7f33-da2e-4ff8-960d-843fdffc9e37.png">

Especially in the linked font size example where font sizes are labelled in 2 letter same-width abbreviations, this works especially well.

However there's an argument to be made that in the Featured Image example shown here, where label widths aren't uniform but buttons are still uniform-width, the uneven space between items is a downside. 

Let me know your thoughts.

## How has this been tested?

Insert the Featured Image block, add a width, then observe the segmented control.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jasmussen committed:
Try: Remove segmented control vertical separators

mtias reviewed (approved):
Looking good to me

ciampo reviewed (commented):
I always like PRs that remove code! Also, as @mtias pointed out, we don't seem to have designs using separators.

My only worry is what @jasmussen also pointed out — that having fixed size buttons with different amounts of text will cause uneven spacing and confuse the user.

We should also be careful and ensure that the space between "toggle buttons" is always large enough to make it clear that it denotes a separation between buttons. Otherwise the user may be getting confused in case a button's label text has spaces.

---

I think that we may go ahead and apply these changes (especially given the fact that this component is still experimental). We can always add the separator back at a later point.

Should we consider giving each `ToggleGroupControlOption` a hover background color to make the separation clearer?

Finally, we should probably add a CHANGELOG entry for this change?

jasmussen commented:
Thanks for the reviews. Looks like I have some snapshots to add.

> I think that we may go ahead and apply these changes (especially given the fact that this component is still experimental). We can always add the separator back at a later point.

I like that, I always like to start simple and add as elements become necessary.

> I think that we may go ahead and apply these changes (especially given the fact that this component is still experimental). We can always add the separator back at a later point.

In fact I think we should add an entry to the component readme that this is a bit of an "editorial responsibility". It's an opinionated opponent, and IMO should never be used if there's ever a chance for the labels, whether in English or another language, to wrap.

jasmussen committed:
Update snaps, add a usage note.

jasmussen commented:
I updated the snapshots, and I added a small comment in the README about usage. I'd love input on whether I did that right, and I'll be sure to follow up. 

> Finally, we should probably add a CHANGELOG entry for this change?

Where would this update go?

ciampo commented:
> Where would this update go?

My bad, I thought I linked it already. It would be [`packages/components/CHANGELOG.md`](https://github.com/WordPress/gutenberg/blob/f37f80fb4ee73b6a30d12063a4479376be137ceb/packages/components/CHANGELOG.md)

jasmussen committed:
Update changelog.

jasmussen commented:
Not at all, and thank you. It's a slow Monday for me — I looked for a changelog section in the README 🙈 

I pushed a change, let me know if that looks good.

ciampo commented on packages/components/CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+### Change
```
Not sure I've seen a `Change` section in the changelog before. Maybe we can call it `Enhancements` for now? 🤷

ciampo commented on packages/components/CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
+### Change
+
+-   Removed the separator shown between items. ([#35497](https://github.com/WordPress/gutenberg/pull/35497)).
```
We can probably be a bit more specific

```suggestion
-   Removed the separator shown between `ToggleGroupControl` items ([#35497](https://github.com/WordPress/gutenberg/pull/35497)).
```

ciampo commented on packages/components/src/toggle-group-control/README.md:
```diff
@@ -6,6 +6,8 @@ This feature is still experimental. “Experimental” means this is an early im
 
 `ToggleGroupControl` is a form component that lets users choose options represented in horizontal segments. To render options for this control use `ToggleGroupControlOption` component.
 
+Only use this control when you know for sure the labels of items inside won't wrap. For items with longer labels, you can consider a Dropdown component instead.
```
The [`Dropdown` component](https://wordpress.github.io/gutenberg/?path=/story/components-dropdown--default) is probably not what you meant here — did you mean [`SelectControl`](https://wordpress.github.io/gutenberg/?path=/story/components-selectcontrol--default) and/or [`CustomSelectControl`](https://wordpress.github.io/gutenberg/?path=/story/components-customselectcontrol--default)?

In any case, if we mention a component's name, it's better to wrap it in backticks for better formatting.

Otherwise, the alternative here could be to be more generic and just mention "you can consider a select/dropdown component instead".

WDYT?

ciampo reviewed (commented):
> It's a slow Monday for me

Haha, I wish my Mondays were "slow" like yours!

Anyway, I left a couple of comments. Would you also mind marking this PR as ready for review? That should potentially enable more CI tasks

jasmussen committed:
Address feedback.

Co-authored-by: @ciampo

jasmussen commented:
Nice, thank you. I pushed some small changes to address your feedback.

jasmussen commented:
@ciampo if you have a second for a last sanity check, then I can land this one. Thanks again.

ciampo reviewed (approved):
LGTM 🚀
